### PR TITLE
support system variable wait_timeout. (#8245)

### PR DESCRIPTION
**Pull request by hhxcc**
### What problem does this PR solve? 
close client connection when the server waited for a long time ,  support system variable `wait_timeout`  to define the wait time.  #8245 

### What is changed and how it works?
use  TCPConn.SetReadDeadline()  before  readPacket()

### Check List 

Tests 
 - Integeration test 

Code changes
 - Has exported variable/fields change

Related changes
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8346)

### CODE REVIEW

**Comment by shenli** [comment]
@hhxcc Thanks! Please add a test case.

File: server/conn.go [hunk 1]
```diff
@@ -210,6 +214,35 @@ func (cc *clientConn) readPacket() ([]byte, error) {
...
+	chanPacketErr := make(chan error, 1)
+	go func() {
+		data, err := cc.readPacket()
+		chanPacketErr <- err
+		chanPacketByte <- data
+		close(chanPacketByte)
+		close(chanPacketErr)
+	}()
+	select {
+	case <-time.After(time.Duration(waitTimeout) * time.Second):
```
Comment (lysu) [inline server/conn.go]: how about using https://golang.org/pkg/net/#TCPConn.SetReadDeadline instead of timer~?

Comment (hhxcc) [reply server/conn.go]: SetReadDeadline is better, I will refactor.

**Comment by hhxcc** [comment]
@shenli  I can only see a litter  test case in conn_test.go,   could you give some advise to write test case ?

**Comment by shenli** [comment]
@lysu Please help @hhxcc to add a test case.

**Comment by hhxcc** [comment]
@lysu  PTAL .  I have do some refactor,  add one test case , bu still do not know how to test the function readPacketWithTimeout.

**Comment by lysu** [comment]
@hhxcc I will try to do it in tomorrow tks~

**Comment by lysu** [comment]
```
func TestConnectionWaitTimeout(t *testing.T) {
 	// this test does not use sql.database, it uses the driver directly
 	if !available {
 		t.Skipf("MySQL server not running on %s", netAddr)
 	}

 	idleDB, err := sql.Open("mysql", dsn)
 	if err != nil {
 		t.Fatalf("error connecting: %s", err.Error())
 	}
 	defer idleDB.Close()
 	_, err = idleDB.Exec("create database if not exists waittimeouttest")
 	assertNoError(t, err)
 	_, err = idleDB.Exec("use waittimeouttest")
 	assertNoError(t, err)

 	checkDB, err := sql.Open("mysql", dsn)
 	if err != nil {
 		t.Fatalf("error connecting: %s", err.Error())
 	}
 	defer checkDB.Close()
 	r := checkDB.QueryRow("select count(1) from information_schema.processlist where db = 'waittimeouttest'")
 	var pc int
 	err = r.Scan(&pc)
 	assertNoError(t, err)
 	assertEqual(t, pc, 1)

 	_, err = idleDB.Exec("set @@wait_timeout=1")
 	assertNoError(t, err)

 	time.Sleep(2*time.Second)

 	r = checkDB.QueryRow("select count(1) from information_schema.processlist where db = 'waittimeouttest'")
 	err = r.Scan(&pc)
 	assertNoError(t, err)
 	assertEqual(t, pc, 0)
 }
```

@hhxcc  I just add a simple testcase in integeration repo which will setup a tidbserver and test them via network.

**Comment by hhxcc** [comment]
@lysu thanks a lot.

File: server/conn.go [hunk 2]
```diff
@@ -447,11 +468,13 @@ func (cc *clientConn) Run() {
...
 		cc.alloc.Reset()
-		data, err := cc.readPacket()
+		data, err := cc.readPacketWithTimeout()
 		if err != nil {
 			if terror.ErrorNotEqual(err, io.EOF) {
 				errStack := errors.ErrorStack(err)
-				if !strings.Contains(errStack, "use of closed network connection") {
+				if strings.Contains(errStack, "i/o timeout") {
+					log.Infof("con:%d read packet timeout, close this connection.", cc.connectionID)
+				} else if !strings.Contains(errStack, "use of closed network connection") {
```
Comment (lysu) [inline server/conn.go]: It's better use this way to detect timeout error

```
		start := time.Now()
		data, err := cc.readPacketWithTimeout()
		idleTime := time.Now().Sub(start)
		if err != nil {
			if terror.ErrorNotEqual(err, io.EOF) {
				if netErr, isNetErr :=  errors.Cause(err).(net.Error); isNetErr && netErr.Timeout() {
					log.Infof("con:%d read packet timeout, close this connection, idle: %v, timeout: %v", cc.connectionID, idleTime, cc.getSessionVarsWaitTimeout())
				} else {
					errStack := errors.ErrorStack(err)
					if !strings.Contains(errStack, "use of closed network connection") {
						log.Errorf("con:%d read packet error, close this connection %s",
							cc.connectionID, errStack)
					}
				}
			}
			return
		}
```

detect close maybe be improve in furture too - - https://github.com/golang/go/issues/4373

**Comment by lysu** [comment]
@hhxcc here, maybe we also take care about a not easy question:

now, deadline is set on `readPacket`, but one `readPacket` maybe trigger several `readOnePacket` then go though the buffer and trigger or not more `read()` call on wire and deadline works on final `read`.

so for a corner case: some one make a big data input to tidb-server, then mysql protocol will transfer them in multiple packet, first packet arrive time is t1, last packet arrive time is t2, and t2 - t1 < wait_time, current PR will give a error.

but I think wait_timeout should be "the timeout that no data on the wire", read deadline should be set for per unbuffed `read()`, just like https://github.com/go-sql-driver/mysql/blob/6be42e0ff99645d7d9626d779001a46e39c5f280/buffer.go#L63 does.

at last, we should take care situation that one conn read by set deadline then read  without deadline, once deadline be set, later read need be reset new deadline or 0(no timeout)

**Comment by hhxcc** [comment]
@lysu  thank you  for  the explanation with so much details,   I will  add `SetReadDeadline` before each `readOnePacket` .
If there are any other problem , please let me know.
thanks again.

**Comment by lysu** [comment]
@hhxcc I think set on each `readOnePacket` fix mainly question, although set on ech `read` is more better, I'm not confirm whelther bufferedConn will introduce other question in some corner case~?

@tiancaiamao @jackysp how do you think about this?

**Comment by hhxcc** [comment]
@lysu   PTAL `SetReadDeadline` has been added before each  `io.ReadFull` ,  and change the way to detect timeout error.

**Comment by lysu** [comment]
LGTM

& 

@jackysp  PTAL thx

File: sessionctx/variable/sysvar.go [hunk 3]
```diff
@@ -535,7 +535,7 @@ var defaultSysVars = []*SysVar{
 	{ScopeGlobal, "innodb_buffer_pool_size", "134217728"},
 	{ScopeGlobal, "innodb_adaptive_flushing", "ON"},
 	{ScopeNone, "datadir", "/usr/local/mysql/data/"},
-	{ScopeGlobal | ScopeSession, "wait_timeout", "28800"},
+	{ScopeGlobal | ScopeSession, WaitTimeout, "28800"},
```
Comment (jackysp) [inline sessionctx/variable/sysvar.go]: It's better to define a const for 28800.

File: sessionctx/variable/sysvar.go [hunk 4]
```diff
@@ -774,6 +774,10 @@ const (
 	SyncBinlog = "sync_binlog"
 	// BlockEncryptionMode is the name for 'block_encryption_mode' system variable.
 	BlockEncryptionMode = "block_encryption_mode"
+	// WaitTimeout is the name for 'wait_timeout' system variable.
+	WaitTimeout = "wait_timeout"
+	// WaitTimeoutDefaultValue is the default value of 'wait_timeout' system variable.
+	WaitTimeoutDefaultValue = "28800"
```
Comment (jackysp) [inline sessionctx/variable/sysvar.go]: Thanks for your quick fix, @hhxcc ! But we usually define the default values of system variable at 
https://github.com/pingcap/tidb/blob/e69aa2749b75d0608ccf062c771c63b737daa05c/sessionctx/variable/tidb_vars.go#L221

Comment (hhxcc) [reply sessionctx/variable/sysvar.go]: ok ,  it has been moved to  tidb_vars.go

**Pull request merged**